### PR TITLE
feat(toolchain): support default toolchain aliases

### PR DESCRIPTION
**Pull request by 14elias**
Related to #5025.
Related discussion: #2793.
this PR implements the **default** special toolcahin name from the proposal in #5025

this change makes **default**  a reserved toolchain alias that always resolves to the currently configured global default toolchain.
It works consistently across all supported override mechanisms:

- Command line: **+default**
- Environment variable: **RUSTUP_TOOLCHAIN=default**
- Directory overrides
- **rust-toolchain.toml** / **rust-toolchain** files

For example, when the configured default is **stable**:
- **cargo +default ...**
- **RUSTUP_TOOLCHAIN=default cargo ...**
- a toolchain file containing **channel = "default"**
all resolves to the **stable**

**Why we need this feature**
The global default has the lowest priority in rustup’s override order.  A project’s **rust-toolchain.toml** therefore completely hides it.  
This change gives users a way to say ignore the project pin and use my personal default right now with out having to hard code the actual name of the default. this change is especially useful in scripts and tools that want to follow the user’s current preference even if that preference later changes.  for example if a script is written like  **cargo +default test**  the script uses what the user's default is with out editing the script.

**Tests**
- +default resolution
- RUSTUP_TOOLCHAIN=default
- rust-toolchain.toml using channel = "default"

### CODE REVIEW

File: src/cli/rustup_mode.rs [hunk 1]
```diff
@@ -929,7 +929,17 @@ async fn default_(
     common::warn_if_host_is_emulated(cfg.process);
 
     if let Some(toolchain) = toolchain {
-        match toolchain.to_owned() {
+        let toolchain = match toolchain.to_owned() {
+            MaybeResolvableToolchainName::Some(ResolvableToolchainName::Default) => {
+                MaybeResolvableToolchainName::Some(
+                    cfg.get_default_resolvable()?
```
Comment (rami3l) [inline src/cli/rustup_mode.rs]: You are repeating this logic too many times to the point where it looks like a more systematic solution is required.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/7df2ba07037827dcc7e0bf7bab4bb8b8fb7b3b26..7b83b06c9c11ed5e7a57ea705a928be32a6ba3b8)*

File: src/cli/rustup_mode.rs [hunk 2]
```diff
@@ -950,6 +960,9 @@ async fn default_(
 
                 common::show_channel_update(cfg, PackageUpdate::Toolchain(desc), Ok(status))?;
             }
+            MaybeResolvableToolchainName::Some(ResolvableToolchainName::Default) => {
+                unreachable!("default should have been resolved above")
```
Comment (rami3l) [inline src/cli/rustup_mode.rs]: We'd like to prevent cases like this if possible, because it means for this type, certain values are no longer valid starting from a certain point in time.

Ideally since we are introducing a new resolution hierarchy, would it be possible to introduce a new type rather than adding a variant to the existing ones?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/7df2ba07037827dcc7e0bf7bab4bb8b8fb7b3b26..7b83b06c9c11ed5e7a57ea705a928be32a6ba3b8)*

**Review by rami3l — CHANGES_REQUESTED** [review]

File: src/cli/rustup_mode.rs (hunk 2 above)
Comment (14elias) [reply src/cli/rustup_mode.rs]: i created new enum type for the alias  and a wrapper to avoid from repeating   the logic                                                                                                           let toolchain = match toolchain.to_owned() {
            MaybeResolvableToolchainName::Some(ResolvableToolchainName::Default) => {
                MaybeResolvableToolchainName::Some(
                    cfg.get_default_resolvable()? 
the wrapper enum has two variants Alias and Explicit  and it has resolve method, so the checking and toolchain dispatching happens in the resolve method.

File: src/cli/self_update.rs [hunk 3]
```diff
@@ -367,6 +367,7 @@ impl InstallOpts<'_> {
                     Some(ResolvableToolchainName::Official(t)) => Some(t),
                     // Default is custom, presumably from a prior install. Do nothing.
                     Some(ResolvableToolchainName::Custom(_)) => None,
+                    // The configured default cannot itself be the default alias.
```
Comment (rami3l) [inline src/cli/self_update.rs]: I think this change is redundant now.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/toolchain/names.rs [hunk 4]
```diff
@@ -74,6 +74,67 @@ pub enum InvalidName {
...
+        if value == "default" {
+            Ok(Self::Default)
+        } else {
+            Err(InvalidName::ToolchainName(value.into()))
+        }
+    }
+}
+
+impl Display for ToolchainAlias {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Comment (rami3l) [inline src/toolchain/names.rs:96]: Suggest importing `std::fmt`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/toolchain/names.rs [hunk 5]
```diff
@@ -74,6 +74,67 @@ pub enum InvalidName {
...
+
+impl FromStr for ToolchainAlias {
+    type Err = InvalidName;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        if value == "default" {
+            Ok(Self::Default)
+        } else {
+            Err(InvalidName::ToolchainName(value.into()))
+        }
```
Comment (rami3l) [inline src/toolchain/names.rs]: Suggest using `match` instead of `if` here so that we can easily add more variants (with exhaustiveness check).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/toolchain/names.rs [hunk 6]
```diff
@@ -74,6 +74,67 @@ pub enum InvalidName {
     DashPrefix(String),
 }
 
+/// An alias for a toolchain name.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ToolchainAlias {
+    Default,
```
Comment (rami3l) [inline src/toolchain/names.rs:81]: Suggest adding a docstring explaining the precise semantics of each variant.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/toolchain/names.rs [hunk 7]
```diff
@@ -74,6 +74,67 @@ pub enum InvalidName {
...
+}
+
+impl<T: FromStr> FromStr for ToolchainOverride<T>
+where
+    T::Err: Into<InvalidName>,
+{
+    type Err = InvalidName;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        if let Ok(candidate) = validate_name(value)
```
Comment (rami3l) [inline src/toolchain/names.rs]: This check makes no sense because it overlaps with the latter check. Note that your aliases are a predetermined set of strings.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/config.rs [hunk 8]
```diff
@@ -669,7 +713,11 @@ impl<'a> Cfg<'a> {
                 None
             };
 
-        Ok(override_config)
+        let Some((override_cfg, source)) = override_config else {
+            return Ok(None);
+        };
+
+        Ok(Some((override_cfg, source)))
```
Comment (rami3l) [inline src/config.rs]: Why is this change necessary?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/config.rs [hunk 9]
```diff
@@ -735,44 +785,49 @@ impl<'a> Cfg<'a> {
...
-                            .list_toolchains(true)?
-                            .iter()
-                            .any(|s| s == resolved_name)
+                        .map_err(|_| {
+                            anyhow!(
+                                "invalid toolchain name detected in override file '{}'",
+                                toolchain_file.display()
+                            )
+                        })?;
+                    if let ToolchainOverride::Explicit(toolchain_name) = toolchain_override {
```
Comment (rami3l) [inline src/config.rs]: I think we may not need to specialize for `Explicit` here, just resolve it as usual (note that `find_override_from_dir_walk` may apply to more than just the project repo; it can totally be a user-local configuration option).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/config.rs [hunk 10]
```diff
@@ -25,12 +25,52 @@ use crate::{
...
+            Self::Alias(ToolchainAlias::Default) => {
+                let default = cfg
+                    .get_default_resolvable()?
+                    .ok_or_else(|| anyhow::anyhow!("no default toolchain is configured"))?;
+                Ok(MaybeResolvableToolchainName::Some(default))
+            }
+            Self::Explicit(r) => Ok(r),
+        }
+    }
+}
```
Comment (rami3l) [inline src/config.rs]: It should be able to replace all this with a single, generic impl block with an appropriate use of the `From` trait. Suggest adding the necessary `From` implementations if needed.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/toolchain/names.rs [hunk 11]
```diff
@@ -74,6 +74,67 @@ pub enum InvalidName {
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Default => write!(f, "default"),
+        }
+    }
+}
+
+/// A wrapper for types that can be overridden by an alias.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ToolchainOverride<T> {
```
Comment (rami3l) [inline src/toolchain/names.rs]: Since the uses of this type are always of the form `ToolchainOverride<xxxToolchain>`, suggest renaming this to just `Override`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/toolchain/names.rs [hunk 12]
```diff
@@ -74,6 +74,67 @@ pub enum InvalidName {
...
+    Alias(ToolchainAlias),
```
Comment (rami3l) [inline src/toolchain/names.rs]: Suggest using the same part of speech (preferably adjective) for all variants. In this case I'd suggest `Aliased` to mirror `Explicit`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/cli/rustup_mode.rs [hunk 13]
```diff
@@ -924,13 +926,15 @@ fn completion_command(cfg: &Cfg<'_>) -> clap::Command {
...
+    toolchain: Option<ToolchainOverride<MaybeResolvableToolchainName>>,
     force_non_host: bool,
 ) -> anyhow::Result<ExitCode> {
     common::warn_if_host_is_emulated(cfg.process);
 
     if let Some(toolchain) = toolchain {
-        match toolchain.to_owned() {
+        let toolchain = toolchain.resolve(cfg)?;
+
+        match toolchain {
```
Comment (rami3l) [inline src/cli/rustup_mode.rs]: Suggest inlining this temporary variable `toolchain`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/cli/proxy_mode.rs [hunk 14]
```diff
@@ -37,10 +37,15 @@ pub async fn main(
     let cfg = Cfg::from_env(current_dir, false, true, process)?;
     let (toolchain, source) = cfg
         .local_toolchain(match toolchain {
-            Some(name) => Some((
-                name.resolve(&cfg.default_host_tuple()?)?,
-                ActiveSource::CommandLine,
-            )),
+            Some(name) => {
+                let name = name.resolve(&cfg)?;
```
Comment (rami3l) [inline src/cli/proxy_mode.rs]: Suggest inlining this temporary variable `name`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

**Review by rami3l — CHANGES_REQUESTED** [review]

File: src/cli/proxy_mode.rs [hunk 15]
```diff
@@ -5,7 +5,7 @@ use crate::{
     command::run_command_for_dir,
```
Comment (rami3l) [inline src/cli/proxy_mode.rs:5]: The category specified in the commit message is wrong. It should look more like `feat(toolchain):` or similar. Same thing for the title of this PR.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

File: src/toolchain/names.rs (hunk 6 above)
Comment (14elias) [reply src/toolchain/names.rs:81]: ///Refers to rustup's configured default toolchain.
///Parsed from the literal string `"default"`.           is this docstring good at representing the variant

Comment (rami3l) [reply src/toolchain/names.rs:81]: @14elias I think just keeping the first line would be fine. Also could you write your comments, PR descriptions etc in proper formatted markdown? It's not super easy to read to me in its current state.

File: src/cli/proxy_mode.rs
Comment (rami3l) [inline src/cli/proxy_mode.rs:2]: Besides, if you have used LLM/agents to draft this PR, please properly disclose the fact in the PR description according to our contribution guidelines. Otherwise, I may close this PR without any prior warnings.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/454ff04cdefebc8f38f47f64b3904866f9e0660f..de81ab2cc212db181d019eaf2c3af2de3a371671)*

Comment (14elias) [reply src/cli/proxy_mode.rs:2]: sorry for the confusion or inconvenience caused, but i did not used AI agent or LLM to write the code, or draft this PR description. i used LLM to brainstorm how to solve the problem specifically after you suggested me to create new data type instead of mixing the alias Default variant to the real tool chain variants and also i used github copilot to check if i applied  ToolchainOverride<T> consistently across the relevant functions and return types and i tested everything by myself, but i will add the tools i used for help  when i update the PR description. thanks for checking.

Comment (14elias) [reply src/cli/proxy_mode.rs:2]: i also used github copilot to help me with tests i mean the role is not like writing the tests it is just like discussing the edge cases and checking what i write and telling me to update the parts i have to modify and what i have to cover. that is how i use AI

File: src/toolchain/names.rs (hunk 7 above)
Comment (14elias) [reply src/toolchain/names.rs]: i kept validate_name() in ToolchainOverride::from_str() because it allow input +default/  as +stable/ . but if aliases should only accept exact names then checking if the value is alias is enough.

File: src/toolchain/names.rs (hunk 4 above)
Comment (14elias) [reply src/toolchain/names.rs:96]: the existing code uses this pattern that's why i used that but i can import std::fmt and change in all the places if you want.

File: src/toolchain/names.rs (hunk 6 above)
Comment (rami3l) [reply src/toolchain/names.rs:84]: Actually you need a space between `///` and the first word.

File: src/toolchain/names.rs (hunk 4 above)
Comment (rami3l) [reply src/toolchain/names.rs]: Yes, please do that. You can import `std::fmt` in a separate commit before this one, and then stack your current changes on top of that one.

File: src/cli/rustup_mode.rs [hunk 16]
```diff
@@ -1794,10 +1796,13 @@ fn pin_active_toolchain(qualified: bool, cfg: &Cfg<'_>) -> anyhow::Result<ExitCo
...
     cfg: &Cfg<'_>,
-    toolchain: ResolvableToolchainName,
+    toolchain: Override<ResolvableToolchainName>,
     path: Option<&Path>,
 ) -> anyhow::Result<ExitCode> {
-    let toolchain_name = toolchain.clone().resolve(&cfg.default_host_tuple()?)?;
+    let resolved_toolchain = toolchain.clone().resolve(cfg)?;
+    let toolchain_name = resolved_toolchain
+        .clone()
+        .resolve(&cfg.default_host_tuple()?)?;
```
Comment (rami3l) [inline src/cli/rustup_mode.rs]: Could you please inline the temporary variable `resolved_toolchain`, or is there anything that prevents you from doing so? Also the `resolved_toolchain.clone()` below may not be necessary.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/7464de1b6121e5ee0fceee3bdcd695ded379cc12..802fa8ef1cf7c6ef35878b171322537e44892b09)*

**Comment by djc** [comment]
Why would people want/need this? What is the use case?

And why add all this complexity with more type system layering, what problem does it solve?

File: src/config.rs [hunk 17]
```diff
@@ -25,12 +25,30 @@ use crate::{
...
+                let default = cfg
+                    .get_default_resolvable()?
+                    .ok_or_else(|| no_toolchain_error(cfg.process))?;
+                Ok(T::from(default))
+            }
+            Self::Explicit(value) => Ok(value),
+        }
+    }
+}
+
```
Comment (rami3l) [inline src/config.rs]: Nit: We'd generally want the `impl` block to live right below the type declaration, so this should be put into `src/toolchain/names.rs` below the declaration of `Override`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/7464de1b6121e5ee0fceee3bdcd695ded379cc12..802fa8ef1cf7c6ef35878b171322537e44892b09)*

File: src/config.rs [hunk 18]
```diff
@@ -668,7 +688,6 @@ impl<'a> Cfg<'a> {
             else {
                 None
             };
-
```
Comment (rami3l) [inline src/config.rs:671]: This whitespace change looks redundant.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/7464de1b6121e5ee0fceee3bdcd695ded379cc12..802fa8ef1cf7c6ef35878b171322537e44892b09)*

**Comment by rami3l** [comment]
> Why would people want/need this? What is the use case?
> 
> And why add all this complexity with more type system layering, what problem does it solve?

@djc This PR implements the `default` layer of https://github.com/rust-lang/rustup/issues/5025. My current plan would be to add only `default` and `file=xxx` layers first, starting with the easiest one (this one, and @14elias happens to have offered to do it), and leave the rest for discussion.

@14elias In the PR description you should have made that clear, otherwise the context may be lost quite easily.

**Comment by djc** [comment]
Still think it's good to explain in this PR when/why people would want this, specifically.

**Review by rami3l — APPROVED** [review]
LGTM, thanks a lot :)

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustup/5071/7464de1b6121e5ee0fceee3bdcd695ded379cc12..90a931710ec868b5439ca9da5861a4ea227bb175)*

**Comment by djc** [comment]
> And why add all this complexity with more type system layering, what problem does it solve?

I'm surprised this was merged without more of a response to these questions.

**Comment by rami3l** [comment]
@djc The author did update the PR description to add more details though. What do you think we should do right now?

**Comment by 14elias** [comment]
> > And why add all this complexity with more type system layering, what problem does it solve?
> 
> I'm surprised this was merged without more of a response to these questions.

i have updated the PR description

**Comment by djc** [comment]
My comment contained two questions: one suggesting a clarification of the context of the PR, and one asking about the complexity of the code in the PR. The first one was adequately handled, but it doesn't feel to me like the second one was considered much.

**Comment by 14elias** [comment]
> My comment contained two questions: one suggesting a clarification of the context of the PR, and one asking about the complexity of the code in the PR. The first one was adequately handled, but it doesn't feel to me like the second one was considered much.

it can be solved simply by checking if the input is default and dispatching it to the configured toolchain but there will be repeating the same codes multiple times specifically the dispatching logic . but this changes follows DRY policy like i write the dispatching in the resolve() method once and i will call it in all the places i want  and also this change allows another aliases to be added easily in the future.  i hope i answered you r question.

**Comment by djc** [comment]
> it can be solved simply by checking if the input is default and dispatching it to the configured toolchain but there will be repeating the same codes multiple times specifically the dispatching logic . but this changes follows DRY policy like i write the dispatching in the resolve() method once and i will call it in all the places i want and also this change allows another aliases to be added easily in the future. i hope i answered you r question.

To be clear, my feedback isn't for you so much as it is for @rami3l.

**Comment by rami3l** [comment]
@djc My apologies for moving too fast here.

That said, I think the wrapper type is useful because override strings are expected to show up in two main cases, `ResolvableToolchainName` and `ResolvableLocalToolchainName`, since different override layers would accept different value types (the difference between the two is because, notably, `RUSTUP_TOOLCHAIN` would accept path-based toolchains).

That said, I dislike the fact that the `none` reserved name is handled in `MaybeOfficialToolchainName` and I'd love to be able to handle all the logic around reserved names in a single place, but I currently don't have a concrete idea for improvement regarding that part. I'd love to work on cleanup PRs if I happen to have found something.

**Comment by djc** [comment]
It feels to me like we have a lot of different "transition" and a whole bunch of supporting code, but we should use a small number of toolchain name types throughout the codebase and proactively resolve to these early/upfront.

It would be good to summarize the design of these and see if some of it can be flattened.
